flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateReduceGroupFunction.scala

```diff
@@ -0,0 +1,179 @@
+  * @param gap Session time window gap.
+  */
+class DataSetSessionWindowAggregateReduceGroupFunction(
+    aggregates: Array[Aggregate[_ <: Any]],
+    groupKeysMapping: Array[(Int, Int)],
+    aggregateMapping: Array[(Int, Int)],
+    intermediateRowArity: Int,
+    finalRowArity: Int,
+    finalRowWindowStartPos: Option[Int],
+    finalRowWindowEndPos: Option[Int],
+    gap:Long)
+  extends RichGroupReduceFunction[Row, Row] {
+
+  private var aggregateBuffer: Row = _
+  private var output: Row = _
+  private var collector: TimeWindowPropertyCollector = _
+  private var intermediateRowWindowStartPos = 0
+  private var intermediateRowWindowEndPos = 0
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
+    Preconditions.checkNotNull(groupKeysMapping)
+    aggregateBuffer = new Row(intermediateRowArity)
+    intermediateRowWindowStartPos = intermediateRowArity - 2
+    intermediateRowWindowEndPos = intermediateRowArity - 1
+    output = new Row(finalRowArity)
+    collector = new TimeWindowPropertyCollector(finalRowWindowStartPos, finalRowWindowEndPos)
+  }
+
+  /**
+    * For grouped intermediate aggregate Rows, divide window according to the window-start
+    * and window-end, merge data (within a unified window) into an aggregate buffer, calculate
+    * aggregated values output from aggregate buffer, and then set them into output
+    * Row based on the mapping relationship between intermediate aggregate data and output data.
+    *
+    * @param records Grouped intermediate aggregate Rows iterator.
+    * @param out     The collector to hand results to.
+    *
+    */
+  override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
+
+    var last: Row = null
+    var head: Row = null
+    var lastWindowEnd: java.lang.Long = null
+    var currentWindowStart:java.lang.Long  = null
+
+    val iterator = records.iterator()
+
+    while (iterator.hasNext) {
+      val record = iterator.next()
+        currentWindowStart = record.getField(intermediateRowWindowStartPos).asInstanceOf[Long]
+        // initial traversal or opening a new window
+        if (null == lastWindowEnd ||
+          (null != lastWindowEnd && currentWindowStart > lastWindowEnd)) {
+
+          // calculate the current window and open a new window
+          if (null != lastWindowEnd) {
+            // evaluate and emit the current window's result.
+            doEvaluateAndCollect(out, last, head)
+          }
+          // initiate intermediate aggregate value.
+          aggregates.foreach(_.initiate(aggregateBuffer))
+          head = record
+        }
+
+        aggregates.foreach(_.merge(record, aggregateBuffer))
+        last = record
+        lastWindowEnd = getWindowEnd(last)
+      }
+
+    doEvaluateAndCollect(out, last, head)
+
+  }
+
+  def doEvaluateAndCollect(
+    out: Collector[Row],
+    last: Row,
+    head: Row): Unit = {
+    // set group keys value to final output.
+    groupKeysMapping.foreach {
```

Comment:
This could be done once in the `reduce()` method using the first record obtained from the iterator (no need to remember `last` for this initialization).